### PR TITLE
Changes to support NAT-PMP 

### PR DESCRIPTION
**Pull request by MishraShivendra**
Changes to support NAT-PMP port forwarding based on libnatpmp to address #11902.
**Changes include the following:**
* Porting of [libnatpmp](https://github.com/miniupnp/libnatpmp) files (namely gateway.cpp/.h and natpmp.cpp/.h)
* C++ style wrapper for easier integration
* Changes to disable upnp by default and use NAT-PMP as explained in #11902 (Introducing -portmap switch).

**Test**
```bash
$./src/bitcoind -portmap
```

Note to reviewers are [here](https://github.com/bitcoin/bitcoin/pull/15717#issuecomment-483001218), need insights.

### CODE REVIEW

**Comment by dongcarl** [comment]
@MishraShivendra I would love to review this, would you be able to do the following:
1. Remove the copyright notices (we're licensed MIT)
2. Break the changes up into smaller, more digestible commits?

**Comment by MishraShivendra** [comment]
Thanks.
Sure. Let me close this pull request and do three separate commits for those bullet points.

**Comment by sipa** [comment]
No need to close the PR; you can overwrite it by force pushing.

**Comment by practicalswift** [comment]
@MishraShivendra Would you mind `clang-format`:ing the new code? https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy

File: src/natpmp/gateway.cpp [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+		if(line > 0) {	/* skip the first line */
+			p = buf;
+			/* skip the interface name */
+			while(*p && !isspace(*p))
+				p++;
+			while(*p && isspace(*p))
+				p++;
+			if(sscanf(p, "%lx%lx", &d, &g)==2) {
+				if(d == 0 && g != 0) { /* default */
+					*addr = g;
```
Comment (practicalswift) [inline src/natpmp/gateway.cpp]: Is the precision losing conversion here intentional? If so, make it explicit :-)

File: src/natpmp/natpmp.cpp [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+}
+
+int SendNewPortMappingRequest(natpmp_t * p, int protocol,
+                              uint16_t privateport, uint16_t publicport,
+							  uint32_t lifetime)
+{
+	if(!p || (protocol!=NATPMP_PROTOCOL_TCP && protocol!=NATPMP_PROTOCOL_UDP))
+		return NATPMP_ERR_INVALIDARGS;
+	p->pending_request[0] = 0;
+	p->pending_request[1] = protocol;
```
Comment (practicalswift) [inline src/natpmp/natpmp.cpp]: Should the valid range for `protocol` be checked here? Perhaps also make the implicit conversion from `int` to `unsigned char` explicit?

File: src/natpmp/natmap_wrapper.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+	fd_set fds;
+	natpmpresp_t response;
+	TimeValType timeout;
+	short retryCount = 0;
+	do {
+		FD_ZERO(&fds);
+		FD_SET(m_NatPmpObj.s, &fds);
+		GetNatPmpRequestTimeout(&m_NatPmpObj, &timeout);
+		returnCode = select(FD_SETSIZE, &fds, NULL, NULL, &timeout);
```
Comment (practicalswift) [inline src/natpmp/natmap_wrapper.cpp]: Use `nullptr´ instead of `NULL`. Applies throughout this PR :-)

File: src/natpmp/natmap_wrapper.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+		}
+		returnCode = ReadNatPmpResponseOrRetry(&m_NatPmpObj, &response);
+		retryCount++;
+	}while(returnCode==NATPMP_TRYAGAIN);
+	
+	if(returnCode<0) {
+		return returnCode;
+	}
+
+	publicPortOut = response.pnu.newportmapping.mappedpublicport,
```
Comment (practicalswift) [inline src/natpmp/natmap_wrapper.cpp]: This looks a bit dangerous: should be `;` and not `,` right? :-)

File: src/natpmp/natpmp.cpp [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+#include <stdio.h>
+
+int InitNatPmp(natpmp_t * p, int forcegw, in_addr_t forcedgw)
+{
+#ifdef WIN32
+	u_long ioctlArg = 1;
+#else
+	int flags;
+#endif
+	struct sockaddr_in addr;
```
Comment (practicalswift) [inline src/natpmp/natpmp.cpp]: Could move `addr` down to first use (`memset`) to make the scope clear and the fact that it is cleared before use.

File: src/natpmp/gateway.cpp [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+#ifndef WIN32
+#include <netinet/in.h>
+#endif
+#if !defined(_MSC_VER)
+#include <sys/param.h>
+#endif
+/* There is no portable method to get the default route gateway.
+ * So below are four (or five ?) differents functions implementing this.
+ * Parsing /proc/net/route is for linux.
+ * sysctl is the way to access such informations on BSD systems.
```
Comment (practicalswift) [inline src/natpmp/gateway.cpp]: Should be "information".

File: src/natpmp/gateway.cpp [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+	//     -[more]
+	//
+	// In order to extract this information, we enumerate each network card, and extract the ServiceName value.
+	// This is then used to open the interface subfolder, and attempt to extract a DhcpDefaultGateway value.
+	// Once one is found, we're done.
+	//
+	// It may be possible to simply enumerate the interface folders until we find one with a DhcpDefaultGateway value.
+	// However, the technique used is the technique most cited on the web, and we assume it to be more correct.
+
+	if(ERROR_SUCCESS != RegOpenKeyEx(HKEY_LOCAL_MACHINE, // Open registry key or predifined key
```
Comment (practicalswift) [inline src/natpmp/gateway.cpp]: Should be "predefined".

File: src/net.cpp [hunk 8]
```diff
@@ -1500,17 +1521,108 @@ void StopMapPort()
...
+		LogPrintf( "Failed to map public port %d to "\
+			   "local port %d with error [%s]\n", publicPort, localPort,
+			   mapper.GetErrMsg(state).c_str());
+		return;
+	}
+	std::thread natMapperReNewer( RenewPortMapping, std::ref(mapper), std::ref(localPort),
+				      std::ref(publicPort), std::ref(life) );
+	natMapperReNewer.detach();
+	g_NatPmpStatus = NatPmpState::MAPPED_PORT;
+	LogPrintf( "Sucessfully mapped local port %d to %s:%d\n",
```
Comment (practicalswift) [inline src/net.cpp]: Should be "Successfully".

File: src/net.cpp [hunk 9]
```diff
@@ -1500,17 +1521,108 @@ void StopMapPort()
...
+	int state = mapper.SendMapReq( NATPMP_PROTOCOL_TCP, localPort, 
+				       publicPort, life);
+	if(state != 0) {
+		LogPrintf( "Failed to remove map between public port %d and "\
+			   "local port %d with error [%s]\n", publicPort, localPort,
+			   mapper.GetErrMsg(state).c_str());
+		return;
+	}
+
+	LogPrintf( "Sucessfully removed map between public port %d and "\
```
Comment (practicalswift) [inline src/net.cpp]: Should be "Successfully".

File: src/natpmp/natmap_wrapper.cpp [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+						   privatePort, publicPortOut,
+						   lifeTimeOut);
+	if(returnCode < 0) {
+		return returnCode;
+	}
+
+	fd_set fds;
+	TimeValType timeout;
+	natpmpresp_t response;
+	short retryCount = 0;
```
Comment (practicalswift) [inline src/natpmp/natmap_wrapper.cpp]: Never used?

File: src/natpmp/natmap_wrapper.cpp [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+	int returnCode = SendPublicAddressRequest(&m_NatPmpObj);
+	if(returnCode<0) {
+		return returnCode;
+	}
+
+
+	fd_set fds;
+	natpmpresp_t response;
+	TimeValType timeout;
+	short retryCount = 0;
```
Comment (practicalswift) [inline src/natpmp/natmap_wrapper.cpp]: Never used?

File: src/natpmp/gateway.cpp [hunk 12]
```diff
@@ -0,0 +1,568 @@
...
+wlan0   0001A8C0        00000000        0001    0       0       0       00FFFFFF        0       0       0
+eth0    0000FEA9        00000000        0001    0       0       0       0000FFFF        0       0       0
+wlan0   00000000        0101A8C0        0003    0       0       0       00000000        0       0       0
+eth0    00000000        00000000        0001    0       0       1000    00000000        0       0       0
+
+ One header line, and then one line by route by route table entry.
+*/
+int GetDefaultGateway(in_addr_t * addr)
+{
+	unsigned long d, g;
```
Comment (practicalswift) [inline src/natpmp/gateway.cpp]: Use fixed with integers type instead of long :-)

Applies throughout this PR.

**Comment by sipa** [comment]
@practicalswift This is code copied from another project; unless we're going to submit improvements upstream, nits here are pointless.

**Comment by practicalswift** [comment]
@MishraShivendra Can you clarify which directories that are meant to be left as-is (with flaws reported upstream) and which directories that are not? :-)

**Comment by sipa** [comment]
@practicalswift It's both in the PR description, and obvious from the code (it's copyright someone else...)

**Comment by practicalswift** [comment]
@sipa The reason that I'm asking explicitly is that I noticed that some files added to this repo were not identical to the corresponding file in the repo that appears to be the upstream.

**Comment by MishraShivendra** [comment]
@practicalswift Thanks for reviewing the changes. Those files were taken from [here]( https://github.com/getlantern/libnatpmp). I've taken just the files needed. A few headers were merged together and useless code was removed. I few other error codes were also added. See diff between corresponding .c and .h files.

> @practicalswift This is code copied from another project; unless we're going to submit improvements upstream, nits here are pointless.

@sipa I think we can fix/modify based on our need. Those files are already way too customized for upstream.

@dongcarl Removing license headers will lead to writing implementation of our own. Not sure if that would be a good idea. Any input?

**Comment by laanwj** [comment]
travis linter complains; either this needs the expected include guards, or to be excluded from the lints
```
src/natpmp/gateway.h seems to be missing the expected include guard:
  #ifndef BITCOIN_NATPMP_GATEWAY_H
  #define BITCOIN_NATPMP_GATEWAY_H
  ...
  #endif // BITCOIN_NATPMP_GATEWAY_H
src/natpmp/natmap_wrapper.h seems to be missing the expected include guard:
  #ifndef BITCOIN_NATPMP_NATMAP_WRAPPER_H
  #define BITCOIN_NATPMP_NATMAP_WRAPPER_H
  ...
  #endif // BITCOIN_NATPMP_NATMAP_WRAPPER_H
src/natpmp/natpmp.h seems to be missing the expected include guard:
  #ifndef BITCOIN_NATPMP_NATPMP_H
  #define BITCOIN_NATPMP_NATPMP_H
  ...
  #endif // BITCOIN_NATPMP_NATPMP_H
^---- failure generated from test/lint/lint-include-guards.sh
```

> Remove the copyright notices (we're licensed MIT)

I don't think this is allowed, at least the 3-clause BSD license has more restrictions than the MIT one.

**Comment by laanwj** [comment]
Ok, we discussed the license issue on IRC.

The conclusion was that it's probably best to mention the differing license for the files in `contrib/debian/copyright` (per file), so that people can see this in one glance without having to view the individual source files, but apart from that it's acceptable to include them. 3BSD versus MIT is close enough, and is not expected to cause problems.

**Comment by dongcarl** [comment]
Also, I believe the upstream is really [here](http://miniupnp.free.fr/files/), I'm not sure if there's a diff between what you copied and what's available upstream, so checking would be great!

**Comment by MarcoFalke** [comment]
>  I believe the upstream is really here

They link to https://github.com/miniupnp/miniupnp in their footer

**Comment by dongcarl** [comment]
I'm sorry about the confusion, https://github.com/miniupnp/libnatpmp is indeed the repo that's being updated.

**Comment by dongcarl** [comment]
@MishraShivendra It seems that Travis is failing. Could you take a look at the errors there?

Aside from that, please rearrange your commits so they look something like this:
1. A commit copying in the newest libnatpmp code
2. A commit integrating libnatpmp with bitcoin's build system
3. One or more commits integrating libnatpmp with bitcoin's codebase

Obviously there might be some mixing of 2 and 3, but at the very least 1 should be a separate commit.

**Comment by luke-jr** [comment]
There's no excuse to subtree this. It should be a normal shared library dependency.

**Comment by laanwj** [comment]
Concept ACK in any case.

**Comment by MishraShivendra** [comment]
@dongcarl @laanwj Thanks for those input and clarification. 

> It seems that Travis is failing. Could you take a look at the errors there?

Changes already in progress on a private branch. I'll push them shortly.

> There's no excuse to subtree this. It should be a normal shared library dependency.

@luke-jr 
I believe it was already discussed in #12288 . Not sure if we need a separate shared library for these changes.?  I'm including 2 files from libnatpmp though.

**Comment by luke-jr** [comment]
There is already a separate shared library.

**Comment by MishraShivendra** [comment]
Travis-ci builds seems to be inconsistent. The above PR build seems to fail but mainline of my forked repo passes: [Travis-Ci-Master-Build](https://travis-ci.org/MishraShivendra/bitcoin/builds/519644709).

@luke-jr Can you please point me to it. I couldn't find any place where we already build one. Did you mean upnp shared lib?

**Comment by practicalswift** [comment]
@MishraShivendra Sure! In light of previous comments -- do you want me to review all the files or just a subset? If so, what subset? :-)

**Comment by MishraShivendra** [comment]
A quick look at everything, please.  
To shed some light on the changes:
[These (first commit)](https://github.com/bitcoin/bitcoin/pull/15717/commits/5293339e4dddc84ca5a219e9b8ca4130b00ef1d8) are the file belong to libnatpmp and [this diff](https://gist.github.com/MishraShivendra/472143f3829cea1026e9fc1714586a85) shows the changes I've introduced in upstream. 
[Second commit](https://github.com/bitcoin/bitcoin/pull/15717/commits/0667cfc11b7c32e48f361d61031e1d53f121e119) introduces integration changes. [net.cpp](https://github.com/bitcoin/bitcoin/pull/15717/files#diff-9a82240fe7dfe86564178691cc57f2f1) includes changes to start/stop/renew the port mapping. RenewPortMapping() is a detached thread to periodically renew the port mapping. Feel free to raise any questions :)

**Comment by luke-jr** [comment]
@MishaShivendra It's installed with the user's OS.

**Comment by MishraShivendra** [comment]
Can someone please take a look at these changes?

File: src/natpmp/gateway.cpp [hunk 13]
```diff
@@ -0,0 +1,548 @@
...
+wlan0   00000000        0101A8C0        0003    0       0       0       00000000        0       0       0
+eth0    00000000        00000000        0001    0       0       1000    00000000        0       0       0
+
+ One header line, and then one line by route by route table entry.
+*/
+
+int TestIfWhiteSpace(char character)
+{
+    std::list<char> whiteSpaces = {'\f', '\n', '\r', '\t', '\v'};
+    return std::count(whiteSpaces.begin(), whiteSpaces.end(), character);
```
Comment (practicalswift) [inline src/natpmp/gateway.cpp]: Take a look at how this is handled in `IsSpace(char c)` (`strencodings.h`).

File: src/natpmp/gateway.cpp [hunk 14]
```diff
@@ -0,0 +1,548 @@
...
+{
+        int64_t d = 0;
+        int64_t g = 0;
+        char buf[256] = {0};
+        int line = 0;
+        char * p = nullptr;
+        FILE* f = fopen("/proc/net/route", "r");
+        if(!f)
+                return FAILED;
+        while(fgets(buf, sizeof(buf), f)) {
```
Comment (practicalswift) [inline src/natpmp/gateway.cpp]: Consider running `clang-tidy` on all new files. For changes to existing files: use https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/clang-format-diff.py

**Comment by MishraShivendra** [comment]
Pushed suggested changes.

File: src/init.cpp [hunk 15]
```diff
@@ -437,16 +437,17 @@ void SetupServerArgs()
     gArgs.AddArg("-torpassword=<pass>", "Tor control port password (default: empty)", false, OptionsCategory::CONNECTION);
 #ifdef USE_UPNP
 #if USE_UPNP
-    gArgs.AddArg("-upnp", "Use UPnP to map the listening port (default: 1 when listening and no -proxy)", false, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-portmap", "Use UPnP to map the listening port (default: 1 when listening and no -proxy)", false, OptionsCategory::CONNECTION);
 #else
-    gArgs.AddArg("-upnp", strprintf("Use UPnP to map the listening port (default: %u)", 0), false, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-portmap", strprintf("Use NAT-PMP to map the listening port (default: %u)", 0), false, OptionsCategory::CONNECTION);
 #endif
```
Comment (luke-jr) [inline src/init.cpp]: Both should be supported...

Comment (MishraShivendra) [reply src/init.cpp]: I meant to use same switch "-portmap" for upnp as well as NAT-PMP. And the particular implementation is picked based on "USE_UPNP" flag (In this file as well as net.cpp). So I believe both are supported but picked based on USE_UPNP flag.

Comment (luke-jr) [reply src/init.cpp]: USE_UPNP is a compile-time option, not run-time.

UPnP and NAT-PMP are two completely different protocols. They don't do the same thing. Some routers may support only one or the other.

Comment (MishraShivendra) [reply src/init.cpp]: @laanwj @luke-jr Do you think we should introduce a separate switch for this? 
#11902

**Comment by MishraShivendra** [comment]
Can someone please give input in the above?

**Comment by luke-jr** [comment]
You should leave UPnP stuff alone for this. Just add NAT-PMP support as a separate feature.

**Comment by ryanofsky** [comment]
There was some discussion about this in IRC yesterday:

http://www.erisian.com.au/bitcoin-core-dev/log-2019-04-24.html#l-255

It does seem like UPnP will be off the table due to concerns about RCE bugs. I did want to throw out as a suggestion that instead of running this code inside bitcoind or even adding it to the project we could consider calling out to a command line client with boost::process like: https://superuser.com/questions/192132/how-to-automatically-forward-a-port-from-the-router-to-a-mac-upnp/192487#192487. This could potentially be a very small code change after #15382 which adds boost::process. But this is just a thought which doesn't need to affect the current effort.

**Comment by dongcarl** [comment]
> There was some discussion about this in IRC yesterday:
> 
> http://www.erisian.com.au/bitcoin-core-dev/log-2019-04-24.html#l-255

Yes that discussion resulted in a few conclusions:

1. NAT-PMP should be enabled by default
2. Let's not submite more reviews of the subtree'd code, as we might make it a dependency under `depends`

w/re how we're going to depend on `libnatpmp`, my naive thinking is that it should be a dependency under `depends` that we statically link into our binaries, since it needs to be on by default. Maybe this doesn't make sense, and others can advise.

**Comment by laanwj** [comment]
> I did want to throw out as a suggestion that instead of running this code inside bitcoind or even adding it to the project we could consider calling out to a command line client with boost::process like: 

Please, no scope creep here. This can always be done later but there's no need to complicate this effort even further with a requirement that it should run outside the process.

**Comment by luke-jr** [comment]
> Updated: We are going to make NAT-PMP part of the bitcoind codebase, exactly like you've done here

NACK. There is no excuse for this. It is universally held to be a bad practice in the open source community, and for very good reasons. We have an excuse to bundle consensus-critical stuff because upstream doesn't generally consider the ramifications of some bugfixes, but that does not apply to NAT-PMP at all.

**Comment by ryanofsky** [comment]
> Please, no scope creep here. 

Just in case there is a misunderstanding, the change I was referring to is potentially a very small change after #15382. Something like:

```c++
gArgs.AddArg("-upnpc_bin", "Path to port forwarding client binary.");
...
if (gArgs.IsArgSet("-upnp_bin")) {
   bp::system(strprintf("%s %i %i", gArgs.GetArg("-upnp_bin"), internal_port, external_port));
}
```

But I mostly just wanted to suggest it an alternative to the approaches discussed above (which all seem perfectly fine to me).

**Comment by Sjors** [comment]
Concept ACK. I'm OK with compiling this by default

I don't think we should turn this on by default until the GUI is updated: that's better done in a followup.

I'm reluctant to strip UPnP compilation by default. The GUI currently has a "map port using UPnP" (off by default). Although it's tempting to just change the text and meaning of that textbox, that's a can of worms by itself. It seems better to just add a checkbox for this new setting, and mention in the tooltip which one is recommended. 

I tend to agree with @luke-jr on avoiding pulling in code. The Depends system seems more appropriate. From IRC in favor of subtree'ing:

> the main reason for not using it as a library was because the upstream library looked to be unmaintained, or at least have no new releases for years

The last [Homebrew release](https://formulae.brew.sh/formula/libnatpmp) is from 2015. Ditto for Ubuntu [Bionic](https://packages.ubuntu.com/bionic/natpmpc). I haven't tested those releases. If they have problems then we could encourage the author to issue a new release. Also, a slight lazy argument, is that anyone who can compile from source can learn how to forward a port; this library primarily benefits users of the binary distribution.

There is still is some [modest commit activity](https://github.com/miniupnp/libnatpmp/commits/master), so we shouldn't declare it dead upstream. Even if it was, I would prefer forking the repo in the bitcoin-core org. The act us of using it could also bring new life to the project. In fact I think we should encourage that, since a lot projects stand to benefit from this library.

For a followup I like the idea of using `boost::process`, but as @laanwj says let's not complicate this PR. It could be a while before #15382 is merged too.

**Comment by MishraShivendra** [comment]
> I did want to throw out as a suggestion that instead of running this code inside bitcoind or even adding it to the project we could consider calling out to a command line client with boost::process like:

I don't seem to agree. It could be difficult to find [these errors](https://github.com/bitcoin/bitcoin/pull/15717/files#diff-7ca56436e371c665a570c330956889caR297) through running the binary with the help of boost::process.

I happen to agree with @luke-jr idea to sub-tree libnatpmp related changes.

Just to sum it up:

1. Leave UPnP as is and keep nat-pmp as a separate feature: Would introduce "-natpmp" switch to enable, and absense of this switch would disable nat-pmp port mapping
2. Sub-tree libnatpmp changes: Would sub-tree libnatpmp. I would remove all unused files and maintain the same gateway/natpmp (.cpp/.h) files introduced already in this PR.

**Comment by luke-jr** [comment]
That's not what I meant. Just use standard `pkg-config` for libnatpmp. Every major distro has it already.

**Comment by Sjors** [comment]
1. Yes
2. What Luke says, plus adding to depends because we need that for distribution builds (though that can wait for a followup if if the feature is off by default)

**Comment by MishraShivendra** [comment]
@luke-jr @Sjors 
So you mean to say #2 should be:
2. Sub-tree libnatpmp changes: Would sub-tree libnatpmp. I would remove all unused files and maintain the same gateway/natpmp (.cpp/.h) files introduced already in this PR. Use package-config for libnatpmp and add depends.

**Comment by Sjors** [comment]
No, I would not sub-tree libnatpmp. Just rely on `pkg-config` to find `libnatpmp` on the system. The wrapper you created can go in `src/util` or some similar place. Don't worry about depends: that can wait for a future commit.

**Comment by brakmic** [comment]
Hi,

First, I'm not sure if this task is still "up for grabs", because it's closed now. However, the icon is still there, so it should be. Anyway, I'm working on it out of curiosity, because I only know about UPnPC, so I thought it would be nice to learn how NAT-PMP works. :)

The [commit in my forked repo](https://github.com/brakmic/bitcoin/commit/df0f5ccb532565d325fad2004f83069dc52f1cc3) contains following changes:

* Integrated **libnatpmp** source directory (similar to secp256k1 and univalue)

* Created *autogen.sh*, *configure.ac* and *Makefile.am* for **libnatpmp**, because the [original repo](https://github.com/miniupnp/libnatpmp) doesn't have them. Instead, they offer a Makefile which makes it non-portable (no .la-files generation available). 

* Added a few **build-aux/m4** macros to search for pthreads, compiler & linker flags, and other stuff.

* Bitcoin's [configure.ac](https://github.com/brakmic/bitcoin/blob/natpmp/configure.ac#L1383) was changed to allow for **searching of system libnatpmp** or **to build it from source**. However, as **libnatpmp** has no pkg-config, the search is not the same as with univalue. Instead, it simply tries to find natpmp.h include file. Not sure, if this is enough for every OS, but we could maybe test it.

* Adapted [src/Makefile.am](https://github.com/brakmic/bitcoin/blob/natpmp/src/Makefile.am#L22) to build and link **libnatpmp**. Also adapted: Makefile.qt.include and Makefile.test.include

* Added [src/natmap.cpp](https://github.com/brakmic/bitcoin/blob/natpmp/src/natmap.cpp) and [src/natmap.h](https://github.com/brakmic/bitcoin/blob/natpmp/src/natmap.cpp) where we access the **libnatpmp** API. The code in these two files is partially based on the example code from [libnatpmp webpage](https://miniupnp.tuxfamily.org/libnatpmp.html) as well as the code from the initial poster of this PR. However, I have removed the whole C-to-C++ mapping stuff, because the C code from libnatpmp is so horrible that there is practically no good way to abstract it away. Instead, I've simply taken all the ugly [only-lowercase](https://github.com/brakmic/bitcoin/blob/natpmp/src/natmap.cpp#L43)-[because-I-hate-humans](https://github.com/brakmic/bitcoin/blob/natpmp/src/natmap.cpp#L59)-Functions and put them directly into NatMap class. [YOLO!](https://github.com/brakmic/bitcoin/blob/natpmp/src/natmap.cpp#L74) Anyway, all kudos go to libnatpmp developer!

Well, that's it. I don't know the current state of this task, but in any case, it was fun writing configure scripts, trying out various Makefile settings,...and reading through ugly C code.

Regards,

**Comment by luke-jr** [comment]
> Integrated libnatpmp source directory (similar to secp256k1 and univalue)

Don't do this.

> non-portable (no .la-files generation available).

.la files are generally considered obsolete...

**Comment by brakmic** [comment]
@luke-jr
I know that la-files are obsolete, but as libnatpmp has no pkg-config, there was no other option. Maybe there is one that I don't know about?

Anyway, I consider this task as obsolete then.

Thanks.

**Comment by Sjors** [comment]
Adding. NAT-PMP support is not obsolete, see #11902. The question is just how.

I think @luke-jr's point about integrating the source directory is that this practice should be avoided, especially for projects that are still being (somewhat) maintained. It's better to make a PR upstream to add pkg-config support.

Meanwhile you could add it to [https://github.com/bitcoin/bitcoin/tree/master/depends], which works by fetching the code from its original source, and has room for patches. The only problem there is that afaik you can't build 1 depends package and get the rest via pkg-config (but I could be wrong).

**Pull request closed without merging**